src/client.ts

```diff
@@ -268,54 +308,71 @@ export class SubscriptionClient {
     };
 
     this.client.onmessage = ({ data }: {data: any}) => {
-      let parsedMessage: any;
-      try {
-        parsedMessage = JSON.parse(data);
-      } catch (e) {
-        throw new Error(`Message must be JSON-parseable. Got: ${data}`);
-      }
-      const subId = parsedMessage.id;
-      if ([KEEPALIVE, INIT_SUCCESS, INIT_FAIL].indexOf(parsedMessage.type) === -1 && !this.subscriptions[subId]) {
-        this.unsubscribe(subId);
+      this.processReceivedData(data);
+    };
+  }
+
+  private processReceivedData(receivedData: any) {
+    let parsedMessage: any;
+    let reqId: number;
+
+    try {
+      parsedMessage = JSON.parse(receivedData);
+      reqId = parsedMessage.id;
+    } catch (e) {
+      throw new Error(`Message must be JSON-parseable. Got: ${receivedData}`);
+    }
+
+    // Complete every request that are not registered on the client but are being sent by the server and are not
+    // in the allowed list
+    if (
+      [ MessageTypes.GQL_DATA,
+        MessageTypes.GQL_COMPLETE,
+        MessageTypes.GQL_ERROR,
+      ].indexOf(parsedMessage.type) !== -1 && !this.requests[reqId]
+    ) {
 
-        if (parsedMessage.type === KEEPALIVE) {
-          return;
+      this.unsubscribe(reqId);
+    }
+
+    switch (parsedMessage.type) {
+      case MessageTypes.GQL_CONNECTION_ERROR:
+        if (this.connectionCallback) {
+          this.connectionCallback(parsedMessage.payload);
         }
-      }
+        break;
 
-      // console.log('MSG', JSON.stringify(parsedMessage, null, 2));
-      switch (parsedMessage.type) {
-        case INIT_FAIL:
-          if (this.connectionCallback) {
-            this.connectionCallback(parsedMessage.payload.error);
-          }
-          break;
-        case INIT_SUCCESS:
-          if (this.connectionCallback) {
-            this.connectionCallback();
-          }
-          break;
-        case SUBSCRIPTION_SUCCESS:
-          delete this.waitingSubscriptions[subId];
-
-          break;
-        case SUBSCRIPTION_FAIL:
-          this.subscriptions[subId].handler(this.formatErrors(parsedMessage.payload.errors), null);
-          delete this.subscriptions[subId];
-          delete this.waitingSubscriptions[subId];
-
-          break;
-        case SUBSCRIPTION_DATA:
-          const payloadData = parsedMessage.payload.data || null;
-          const payloadErrors = parsedMessage.payload.errors ? this.formatErrors(parsedMessage.payload.errors) : null;
-          this.subscriptions[subId].handler(payloadErrors, payloadData);
-          break;
-        case KEEPALIVE:
-          break;
-
-        default:
-          throw new Error('Invalid message type!');
-      }
-    };
+      case MessageTypes.GQL_CONNECTION_ACK:
+        if (this.connectionCallback) {
+          this.connectionCallback();
```

Comment:
I think this is something we were doing for subscriptions, right? Maybe I even implemented this the first time around, but I don't like it that much any more 😬  It's kind of overloading the callback. Do we really need to know when it was connected? I might prefer exposing that in a `connectionStatus()` that can be called from outside, or a `onConnectionStatusChange()` callback. That would feel a lot less hacky and be much more versatile.